File: BOJ/DP/1932.cpp
// 1932 정수 삼각형
// https://www.acmicpc.net/problem/1932
#include <iostream>
#include <vector>

using namespace std;

int mx(int a, int b) {
    return a > b ? a : b;
}

int solution(int N, vector<vector<int>> board) {
    int answer = 0;
    vector<vector<int>> dp(N, vector<int> (N, 0));

    dp[0][0] = board[0][0];

    for (int i = 1; i < N; ++i)
    {
        for (int j = 0; j <= i; ++j)
        {
            dp[i][j] = board[i][j]+mx(dp[i-1][j], dp[i-1][j-1]);
        }
    }

    for (int i = 0; i < N; ++i)
    {
        answer = mx(dp[N-1][i], answer);
    }

    return answer;
}

int main(void) {
    int N;
    vector<vector<int>> board;

    cin >> N;
    board = vector<vector<int>> (N, vector<int> (N, -1));
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j <= i; ++j)
        {
            cin >> board[i][j];
        }
    }

    cout << solution(N, board);

    return 0;
}

File: BOJ/DP/1932.py
# 1932 정수 삼각형
# https://www.acmicpc.net/problem/1932
import sys
input = sys.stdin.readline


def solution(n, triangle):
    dp = [[0] * (i+1) for i in range(n)]

    dp[0][0] = triangle[0][0]

    for i in range(1, n):
        for j in range(i+1):
            if j == 0:
                dp[i][j] = triangle[i][j] + dp[i-1][j]
            elif j == i:
                dp[i][j] = triangle[i][j] + dp[i-1][j-1]
            else:
                dp[i][j] = triangle[i][j] + max(dp[i-1][j], dp[i-1][j-1])

    return max(dp[n-1])


n = int(input())
triangle = [list(map(int, input().split())) for _ in range(n)]

print(solution(n, triangle))
